shell/common/pointer_data_dispatcher.cc

```diff
@@ -60,4 +61,170 @@ void SmoothPointerDataDispatcher::DispatchPendingPacket() {
+
+void ResamplingPointerDataDispatcher::DispatchPacket(
+    std::unique_ptr<PointerDataPacket> packet,
+    uint64_t trace_flow_id) {
+  TRACE_EVENT0("flutter", "ResamplingPointerDataDispatcher::DispatchPacket");
+  size_t kBytesPerPointerData = kPointerDataFieldCount * kBytesPerField;
+  auto buffer = packet->data();
+  size_t buffer_length = buffer.size();
+  for (size_t i = 0; i < buffer_length / kBytesPerPointerData; i++) {
+    PointerData data;
+    memcpy(&data, &buffer[i * kBytesPerPointerData], sizeof(PointerData));
+    auto trace_arg = std::to_string(data.time_stamp);
+    TRACE_EVENT1("flutter", "PointerEvent", "time_stamp", trace_arg.c_str());
+    PointerEvent event;
+    event.data = data;
+    event.trace_flow_id = trace_flow_id;
+    pending_events_.push_back(event);
+  }
+  TRACE_FLOW_STEP("flutter", "PointerEvent", trace_flow_id);
+  ScheduleSecondaryVsyncCallback();
+}
+
+void ResamplingPointerDataDispatcher::ScheduleSecondaryVsyncCallback() {
+  if (secondary_vsync_callback_pending_) {
+    return;
+  }
+  secondary_vsync_callback_pending_ = true;
+  delegate_.ScheduleSecondaryVsyncCallback(
+      [dispatcher = weak_factory_.GetWeakPtr()](
+          fml::TimePoint frame_start_time, fml::TimePoint frame_target_time) {
+        if (dispatcher) {
+          auto trace_arg =
+              std::to_string(frame_start_time.ToEpochDelta().ToMicroseconds());
+          TRACE_EVENT1("flutter",
+                       "ResamplingPointerDataDispatcher::VsyncCallback",
+                       "frame_start_time_us", trace_arg.c_str());
+          dispatcher->sample_time_us_ =
+              frame_start_time.ToEpochDelta().ToMicroseconds() +
+              dispatcher->sampling_offset_us_;
+          dispatcher->secondary_vsync_callback_pending_ = false;
+          dispatcher->DispatchPendingPackets();
+        }
+      });
+}
+
+void ResamplingPointerDataDispatcher::DispatchPendingPackets() {
+  TRACE_EVENT0("flutter",
+               "ResamplingPointerDataDispatcher::DispatchPendingPackets");
+
+  // Consume pending events.
+  while (!pending_events_.empty()) {
+    auto& event = pending_events_.front();
+
+    // Stop consuming events if more recent than current sample time.
+    if (event.data.time_stamp > sample_time_us_) {
+      break;
+    }
+
+    int device = event.data.device;
+    switch (event.data.change) {
+      case flutter::PointerData::Change::kDown: {
+        DownPointer pointer;
+        pointer.last = event;
+        pointer.next = event;
+        down_pointers_[device] = pointer;
+      } break;
+      case flutter::PointerData::Change::kMove:
+        down_pointers_[device].next = event;
+        break;
+      case flutter::PointerData::Change::kCancel:
+      case flutter::PointerData::Change::kUp:
+        down_pointers_.erase(device);
+        break;
+      case flutter::PointerData::Change::kAdd:
+      case flutter::PointerData::Change::kRemove:
+      case flutter::PointerData::Change::kHover:
```

Comment:
What does "resampling" mean in this context? I'm not sure what the question is exactly.

(I'm also confused because unless I'm misremembering `kHover`->`kMove` isn't a valid transition.)

Reply:
Resampling here means to send artificial move events that are in sync with the vsync signal, even if the input device is sampled at random times not in sync with vsync.

It's nice to know that `kHover -> kMove` isn't valid. Is `kDown -> kMove` the only valid transition to `kMove`? In general, if any `kMove` affects UI rendering, we'd like those `kMove` events to be resampled and in sync with vsync. Otherwise, the rendering may look janky.

Reply:
We should have clear documentation somewhere about the exact allowed state transitions. If we don't, the first step would be to document that.

Reply:
@Hixie There's documentation about the phases in embedder.h, but it's not a rigorous state transition model. Although when I say it's not valid, that may be one of the cases where there's fixup happening on the Dart side to try to smooth out discrepancies in the various platform implementations, so it may be "valid" at the embedder API level even though it shouldn't be. (I know there were discussions about potentially doing that kind of fixup in the embeddings rather than Dart, which seemed like a good idea to me, but IIRC that was abandoned. If we want to rigorously define the allowed transitions we should probably make sure we're happy with the structure we have first.)

@liyuqian Currently we don't do anything like what you are describing on desktop. The only synthetic events we ever send are to clean up discrepancies in the OS events in some edge cases.

Reply:
> @liyuqian Currently we don't do anything like what you are describing on desktop. The only synthetic events we ever send are to clean up discrepancies in the OS events in some edge cases.

@stuartmorgan : right, currently there isn't resampling in any Flutter platforms. I'm asking here to figure out what's the best way to resample for all platforms, including the desktop.

Reply:
My understanding was that we were moving to a world where we have a clear set of rules that the events must follow, and that the platform-specific code was responsible for ensuring the events from the platform were coerced to follow that contract. I don't know what the status of that is, but that's where we should be headed.